File: Python/watertank_a_bfs.py
import collections
import sys

# Getting parameters
vol_wanted = int(sys.argv[3])
containers_count = 2
containers_volume = [int(sys.argv[1]), int(sys.argv[2])]
check_dict = {}
starting_node = [[0, 0]]


# GCD function needed to check if there is a solution
def gcd(a, b):
    if b == 0:
        return a
    return gcd(b, a % b)


def check_args(cont_vol, vol_wanted):
    """
        If water required is superior of the biggest tank, there's no solution
        With the Bezout Lemma we can check if there's a solution such as + by = z
        Lemma : if x and y are nonzero integers and g = gcd(x,y),
        then there exist integers a and b such that ax+by=g.
    """
    if vol_wanted > max(cont_vol[0], cont_vol[1]) or (vol_wanted % (gcd(cont_vol[0], cont_vol[1])) is not 0):
        print("No solution")
        return False
    return True


def search(starting_node, containers_volume, vol_wanted, check_dict):
    # Starting state
    target = []
    accomplished = False

    # Using collections and deque to have a list with improved method appendleft, popleft etc..
    q = collections.deque()
    q.appendleft(starting_node)

    if check_args(containers_volume, vol_wanted):
        while len(q) != 0:
            path = q.popleft()
            check_dict[get_index(path[-1])] = True

            # print other path
            # if len(path) >= 2:
            #    print(transition(path[-2], path[-1], containers_volume), path[-1])

            # If target is reached, exit the loop
            if is_vol_wanted(path, vol_wanted):
                accomplished = True
                target = path
                break

            # If target no reached, getting the next move
            next_moves = next_step(containers_volume, path, check_dict)
            for i in next_moves:
                q.append(i)

        # If accomplished, print the path
        if accomplished:
            print_path(target)
        else:
            print("No Solution")
            return


def been_there(node, check_dict):
    # Check if a node had already been visited

    return check_dict.get(get_index(node), False)


def next_step(containers_volume, path, check_dict):
    # Finding the next path with checking the ones already checked

    result = []
    next_nodes = []
    node = []

    max_first_cont = containers_volume[0]
    max_second_cont = containers_volume[1]

    first_cont = path[-1][0]  # initial amount in the first tank
    second_cont = path[-1][1]  # initial amount in the second tank

    # fill in the first tank
    node.append(max_first_cont)
    node.append(second_cont)
    if not been_there(node, check_dict):
        next_nodes.append(node)
    node = []

    # fill in the second tank
    node.append(first_cont)
    node.append(max_second_cont)
    if not been_there(node, check_dict):
        next_nodes.append(node)
    node = []

    #  Transfer second tank to first tank (B => A)
    node.append(min(max_first_cont, first_cont + second_cont))
    node.append(second_cont - (node[0] - first_cont))
    if not been_there(node, check_dict):
        next_nodes.append(node)
    node = []

    ##  Transfer first tank to second tank (A => B)
    node.append(min(first_cont + second_cont, max_second_cont))
    node.insert(0, first_cont - (node[0] - second_cont))
    if not been_there(node, check_dict):
        next_nodes.append(node)
    node = []

    # Empty the first tank
    node.append(0)
    node.append(second_cont)
    if not been_there(node, check_dict):
        next_nodes.append(node)
    node = []

    # Empty the second tank
    node.append(first_cont)
    node.append(0)
    if not been_there(node, check_dict):
        next_nodes.append(node)

    # Create the list of possible next path
    for i in range(0, len(next_nodes)):
        temp = list(path)
        temp.append(next_nodes[i])
        result.append(temp)

    return result


def write_transition(old, new):
    first_cont = old[0]
    second_cont = old[1]
    new_first_cont = new[0]
    new_second_cont = new[1]

    if first_cont > new_first_cont:
        if second_cont == new_second_cont:
            return f'A->* :'
        else:
            return f'A->B :'
    else:
        if second_cont > new_second_cont:
            if first_cont == new_first_cont:
                return f'B->* :'
            else:
                return f'B->A :'
        else:
            if first_cont == new_first_cont:
                return f'*->B :'
            else:
                return f'*->A :'


def print_path(path):
    print(f'Solution in {len(path) - 1} steps')
    for i in range(0, len(path) - 1):
        # print(i + 1, ":", transition(path[i], path[i + 1], containers_volume), path[i + 1])
        print(write_transition(path[i], path[i + 1]), tuple(path[i + 1]))


def is_vol_wanted(path, vol_wanted):
    # Checking the last node and return True if we reached the targeted value
    return path[-1][0] == vol_wanted or path[-1][1] == vol_wanted


def get_index(node):
    # generating a random number for the key index using " ** " to be sure there's no duplicate
    return 3 ** node[0] * 4 ** node[1]


search(starting_node, containers_volume, vol_wanted, check_dict)

File: Java/src/Watertank_a_bfs.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import javax.lang.model.type.ArrayType;
import java.lang.reflect.Array;
import java.util.*;

// dynamic programming method using Breadth-First Search (BFS) Algorithm

public class Watertank_a_bfs {

    public static void main(String[] args) {
        // General Parameters
        ArrayList<Integer> containers_volume = new ArrayList<Integer>();
        ArrayList<ArrayList<String>> steps = new ArrayList<ArrayList<String>>();

        containers_volume.add(Integer.parseInt(args[0]));
        containers_volume.add(Integer.parseInt(args[1]));
        int vol_wanted=Integer.parseInt(args[2]);

        ArrayList starting_node=new ArrayList<>();
        ArrayList <Integer> starting_node_temp=new ArrayList<>();
        starting_node_temp.add(0);
        starting_node_temp.add(0);
        starting_node= starting_node_temp;
        HashMap check_dict = new HashMap();


        search(starting_node, containers_volume, vol_wanted, check_dict);

    }

    // # GCD function needed to check if there is a solution
    public static int gcd(int a, int b)
    {
        if(b==0)
        {
            return a;
        }
        else
        {
            return gcd(b,a%b) ;
        }
    }

    public static boolean check_arg(ArrayList<Integer> containers_volume, int vol_wanted)
    {
        // If water required is superior of the biggest tank OR Bezout lemma is not respected, it's false (if x and y are nonzero integers and g = gcd(x,y),
        // then there exist integers a and b such that ax+by=g.
        if( (vol_wanted> Math.max( (int) containers_volume.get(0), (int) containers_volume.get(1) )) || (vol_wanted % (gcd((int) containers_volume.get(0), (int) containers_volume.get(1))) !=0) )
        {
            System.out.println("No solution");
            return false;
        }
        else
        {
            return true;
        }

    }


    public static void search(ArrayList starting_node, ArrayList containers_volume, int vol_wanted, HashMap check_dict)
    {
        ArrayList<Integer> target=new ArrayList<>();
        ArrayList<ArrayList<Integer>> q =new ArrayList<>();
        ArrayList path = new ArrayList<>();
        ArrayList path_2;
        ArrayList<ArrayList<Integer>> next_moves = new ArrayList<>();
        boolean accomplished= false;

        q.add(starting_node);

        if(check_arg(containers_volume, vol_wanted))
        {
            while(q.size() != 0) {
                path_2 = (ArrayList) q.get(0);
                ArrayList comp = new ArrayList();
                comp.add(0);
                comp.add(0);
                ArrayList comp_2 = comp;
                if (path_2.equals(comp_2))
                {
                    path.add(new ArrayList(path_2));
                }
                else{
                    path = (new ArrayList(path_2));
                }
                q.remove(0);
                ArrayList last_path = new ArrayList();
                last_path = (ArrayList) path.get(path.size() - 1);
                double b = get_index(last_path);
                check_dict.put(b, true);

                // If target is reached, exit the loop
                if (is_vol_wanted(path, vol_wanted)) {
                    accomplished = true;
                    target = path;
                    break;
                }

                // If target no reached, getting the next move
                next_moves = next_step(containers_volume, path, check_dict);
                for (int i = 0; i < next_moves.size(); i++) {
                    q.add(next_moves.get(i));

                }
            }
            // If accomplished, print the path
            if(accomplished)
            {
                print_path(target);
            }
            else
            {
                System.out.println("no solution");
            }
        }
    }





    public static boolean been_there(ArrayList node, HashMap check_dict)
            // Check if a node had already been visited
    {
        double a=get_index(node);
        if (check_dict.containsKey(a))
        {
            return true;
        }
        else {
            return false;
        }
    }

    public static ArrayList next_step(ArrayList containers_volume, ArrayList path, HashMap check_dict)
            // Finding the next path with checking the ones already checked
    {
        ArrayList result=new ArrayList<>();
        ArrayList<ArrayList> next_node=new ArrayList<>();
        ArrayList<Integer> node=new ArrayList<>();
        ArrayList last_path;

        int max_first_cont= (int) containers_volume.get(0);
        int max_second_cont= (int) containers_volume.get(1);


        last_path = (ArrayList) path.get(path.size()-1);
        int first_cont=(int) last_path.get(0);  // initial amount in the first tank
        int second_cont=(int) last_path.get(1);  // initial amount in the second tank


        // fill in the first tank
        node.add(max_first_cont);
        node.add(second_cont);
        if(!been_there(node, check_dict)){
            next_node.add(new ArrayList(node));
        }
        node.clear();

        // fill in the second tank
        node.add(first_cont);
        node.add(max_second_cont);
        if(!been_there(node, check_dict)){
            next_node.add(new ArrayList(node));

        }
        node.clear();

        // Transfer second tank to first tank (B => A)
        node.add(Math.min(max_first_cont, first_cont + second_cont));
        node.add(second_cont-((int) node.get(0)-first_cont));
        if(!been_there(node, check_dict))
        {
            next_node.add(new ArrayList(node));
        }
        node.clear();

        // Transfer first tank to second tank (A => B)
        node.add(Math.min(first_cont + second_cont, max_second_cont));
        node.add(0,second_cont-((int) node.get(0)-first_cont));
        if(!been_there(node, check_dict))
        {
            next_node.add(new ArrayList(node));
        }
        node.clear();

        // Empty the first tank
        node.add(0);
        node.add(second_cont);
        if(!been_there(node,check_dict))
        {
            next_node.add(new ArrayList(node));
        }
        node.clear();

        // Empty the second tank
        node.add(first_cont);
        node.add(0);
        if(!been_there(node, check_dict)){
            next_node.add(new ArrayList(node));
        }

        if(next_node.size() != 0);
        {
            // Create the list of possible next path
            for(int i=0; i<next_node.size(); i++)
            {
                //temp=path;
                ArrayList temp = new ArrayList();
                temp = new ArrayList(path);
                temp.add(next_node.get(i));
                result.add(new ArrayList(temp));
                temp.removeAll(result);
            }

        }
        return result;

    }

    public static void write_transition(ArrayList old, ArrayList m_new)
    {
        int first_cont=(int) old.get(0);
        int second_cont=(int) old.get(1);
        int new_first_cont=(int) m_new.get(0);
        int new_second_cont=(int) m_new.get(1);

        if(first_cont>new_first_cont){
            if(second_cont==new_second_cont)
            {
                System.out.print("A->* : ");
            }
            else
            {
                System.out.print("A->B : ");
            }

        }
        else
        {
            if (second_cont > new_second_cont) {
                if (first_cont == new_first_cont){
                    System.out.print("B->* : ");
                }
                else
                {
                    System.out.print("B->A : ");
                }
            }
            else
            {
                if(first_cont == new_first_cont)
                {
                    System.out.print("*->B : ");
                }
                else
                {
                    System.out.print("*->A : ");
                }
            }
        }

    }

    public static  void print_path(ArrayList path)
    {
        System.out.println("Solution in "+(path.size()-1)+" steps");
        for(int i =0; i<path.size()-1; i++)
        {
            write_transition((ArrayList) path.get(i), (ArrayList) path.get(i+1));
            String step = new String();
            step = path.get(i+1).toString().replaceAll("\\[","\\(").replaceAll("\\]","\\)");
            System.out.println(step);
        }
    }

    public static boolean is_vol_wanted(ArrayList path, int vol_wanted)
    {
        //Checking the last node and return True if we reached the targeted value

        Integer temp4=0;
        Integer temp5=0;
        ArrayList last_path = new ArrayList();
        last_path = (ArrayList) path.get(path.size()-1);
        temp4= (int) last_path.get(0);
        temp5= (int) last_path.get(1);

        return temp4==vol_wanted || temp5==vol_wanted;

    }

    public static double get_index(ArrayList node)
    {
        // generating a random number for the key index using pow to be sure there's no duplicate
        return Math.pow(3,(int) node.get(0)) * Math.pow(4,(int) node.get(1));
    }

}



